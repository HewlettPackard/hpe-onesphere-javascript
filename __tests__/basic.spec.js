// (C) Copyright 2017-2018 Hewlett Packard Enterprise Development LP
import OneSphere from '../src';

const URL = process.env.ONESPHERE_URL;
const USERNAME = process.env.ONESPHERE_USERNAME;
const PASSWORD = process.env.ONESPHERE_PASSWORD;

if (!URL) {
  console.error('Missing ONESPHERE_URL environment variable.');
}
if (!USERNAME) {
  console.error('Missing ONESPHERE_USERNAME environment variable.');
}
if (!PASSWORD) {
  console.error('Missing ONESPHERE_PASSWORD environment variable.');
}
if (!URL || !USERNAME || !PASSWORD) {
  process.exit(1);
}

describe('basic', () => {
  let oneSphere;

  beforeAll((done) => {
    oneSphere = new OneSphere(URL);
    oneSphere.postSession({ username: USERNAME, password: PASSWORD })
      .then(() => done());
  });

  test('Status', () =>
    expect(oneSphere.getStatus()
      .then(status => Object.keys(status)))
      .resolves.toMatchSnapshot());

  test('Session', () =>
    expect(oneSphere.getSession()
      .then(session => Object.keys(session)))
      .resolves.toMatchSnapshot());

  test('User', () =>
    expect(oneSphere.getUsers()
      .then(users => Object.keys(users.members[0])))
      .resolves.toMatchSnapshot());

  test('Role', () =>
    expect(oneSphere.getRoles()
      .then(roles => Object.keys(roles.members[0])))
      .resolves.toMatchSnapshot());

  test('Catalog Types', () =>
    expect(oneSphere.getCatalogTypes()
      .then(catalogTypes => Object.keys(catalogTypes.members[0])))
      .resolves.toMatchSnapshot());

  test('Catalog', () =>
    expect(oneSphere.getCatalogs()
      .then(catalogs => Object.keys(catalogs.members[0])))
      .resolves.toMatchSnapshot());

  test('ServiceType', () =>
    expect(oneSphere.getServiceTypes())
      .resolves.toMatchSnapshot());

  test('Service', () =>
    expect(oneSphere.getServices()
      .then(services => Object.keys(services.members[0])))
      .resolves.toMatchSnapshot());

  test('VirtualMachineProfile', () =>
    expect(oneSphere.getVirtualMachineProfiles()
      .then(vmProfiles => Object.keys(vmProfiles.members[0])))
      .resolves.toMatchSnapshot());

  test('TagKey', () =>
    expect(oneSphere.getTagKeys()
      .then(tagKeys => Object.keys(tagKeys.members[0])))
      .resolves.toMatchSnapshot());

  test('Tag', () =>
    expect(oneSphere.getTags()
      .then(tags => Object.keys(tags.members[0])))
      .resolves.toMatchSnapshot());

  test('ProviderType', () =>
    expect(oneSphere.getProviderTypes())
      .resolves.toMatchSnapshot());

  test('Provider', () =>
    expect(oneSphere.getProviders()
      .then(providers => Object.keys(providers.members[0])))
      .resolves.toMatchSnapshot());

  test('Region', () =>
    expect(oneSphere.getRegions()
      .then(regions => Object.keys(regions.members[0])))
      .resolves.toMatchSnapshot());

  test('ZoneType', () =>
    expect(oneSphere.getZoneTypes())
      .resolves.toMatchSnapshot());

  test('Zone', () =>
    expect(oneSphere.getZones()
      .then(zones => Object.keys(zones.members[0])))
      .resolves.toMatchSnapshot());

  test('Rate', () =>
    expect(oneSphere.getRates()
      .then(rates => Object.keys(rates.members[0])))
      .resolves.toMatchSnapshot());

  test('Project', () =>
    expect(oneSphere.getProjects()
      .then(projects => Object.keys(projects.members[0])))
      .resolves.toMatchSnapshot());

  test('Deployment', () =>
    expect(oneSphere.getDeployments()
      .then(deployments => Object.keys(deployments.members[0])))
      .resolves.toMatchSnapshot());

  test('Metric', () =>
    expect(oneSphere.getMetrics({ category: 'projects', period: 'month', name: 'cost.usage' })
      .then(metrics => Object.keys(metrics.members[0])))
      .resolves.toMatchSnapshot());
});

describe('add, update, delete', () => {
  let oneSphere;

  beforeAll((done) => {
    oneSphere = new OneSphere(URL);
    oneSphere.postSession({ username: USERNAME, password: PASSWORD })
      .then(() => done());
  });

  // requires being 'analyst' and not only 'user'
  // => should it be in a separate file analyst.spec.js?

  test('TagKey', (done) => {
    const data = {
      'name': 'Api Test TagKey',
    };
    // add project
    oneSphere.addTagKey(data)
      .then((tagkey) => {
        expect(Object.keys(tagkey)).toMatchSnapshot();
        expect(tagkey).toMatchObject(expect.objectContaining(data));
        return tagkey;
      })

      // get tagkey
      .then(tagkey => oneSphere.getTagKey(tagkey.uri))
      .then((tagkey) => {
        expect(Object.keys(tagkey)).toMatchSnapshot();
        expect(tagkey).toMatchObject(expect.objectContaining(data));
        return tagkey;
      })

      // no PATCH request to update tagkey availables

      // remove tagkey
      .then(tagkey => oneSphere.removeTagKey(tagkey.uri))
      .then(() => done());
  }, 5000); // 5s empirically determined
});
