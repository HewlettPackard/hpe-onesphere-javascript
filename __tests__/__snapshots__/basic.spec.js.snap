// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add, update, delete TagKey 1`] = `
Array [
  "id",
  "name",
  "uri",
]
`;

exports[`add, update, delete TagKey 2`] = `
Array [
  "id",
  "name",
  "uri",
]
`;

exports[`basic Catalog 1`] = `
Array [
  "id",
  "name",
  "uri",
  "serviceTypeUri",
  "catalogTypeUri",
  "zoneUri",
  "url",
  "status",
  "state",
  "protected",
  "created",
  "modified",
]
`;

exports[`basic Catalog Types 1`] = `
Array [
  "id",
  "name",
  "uri",
  "protected",
]
`;

exports[`basic Deployment 1`] = `
Array [
  "id",
  "name",
  "zoneUri",
  "regionUri",
  "serviceUri",
  "serviceTypeUri",
  "virtualMachineProfileUri",
  "status",
  "state",
  "memorySizeGB",
  "cpuCount",
  "cpuGhz",
  "diskSizeGB",
  "uri",
  "projectUri",
  "deploymentEndpoints",
  "hasConsole",
  "created",
  "modified",
]
`;

exports[`basic Metric 1`] = `
Array [
  "name",
  "resourceUri",
  "category",
  "units",
  "description",
  "associations",
  "values",
]
`;

exports[`basic Project 1`] = `
Array [
  "id",
  "name",
  "description",
  "uri",
  "modified",
  "tagUris",
  "protected",
]
`;

exports[`basic Provider 1`] = `
Array [
  "id",
  "name",
  "uri",
  "providerTypeUri",
  "status",
  "state",
  "projectUris",
  "paymentProvider",
  "s3CostBucket",
  "created",
  "modified",
]
`;

exports[`basic ProviderType 1`] = `
Object {
  "count": 2,
  "members": Array [
    Object {
      "id": "aws",
      "logo": "https://d2p6eldpu75wyk.cloudfront.net/provider-thumb-amazon-web-services.svg",
      "logoType": "https://d2p6eldpu75wyk.cloudfront.net/provider-logo-amazon-web-services.svg",
      "name": "Amazon Web Services",
      "uri": "/rest/provider-types/aws",
    },
    Object {
      "id": "ncs",
      "logo": "https://d2p6eldpu75wyk.cloudfront.net/provider-thumb-hpe-new-stack.svg",
      "logoType": "https://d2p6eldpu75wyk.cloudfront.net/provider-logo-hpe-new-stack.svg",
      "name": "Private Cloud",
      "uri": "/rest/provider-types/ncs",
    },
  ],
  "start": 0,
  "total": 2,
}
`;

exports[`basic Rate 1`] = `
Array [
  "rateUri",
  "resourceUri",
  "resourceTerminated",
  "metricName",
  "name",
  "effectiveDate",
  "rateValue",
  "unit",
  "created",
  "deleted",
  "active",
]
`;

exports[`basic Region 1`] = `
Array [
  "id",
  "name",
  "uri",
  "location",
  "providerUri",
  "status",
  "state",
  "created",
  "modified",
]
`;

exports[`basic Role 1`] = `
Array [
  "id",
  "name",
  "displayName",
  "uri",
]
`;

exports[`basic Service 1`] = `
Array [
  "id",
  "name",
  "description",
  "detailedDescription",
  "detailedDescriptionURL",
  "regionUri",
  "zoneUri",
  "projectUri",
  "serviceTypeUri",
  "uri",
  "icon",
  "catalogUri",
  "versions",
  "created",
  "modified",
]
`;

exports[`basic ServiceType 1`] = `
Object {
  "count": 4,
  "members": Array [
    Object {
      "id": "chart",
      "name": "Helm Chart",
      "uri": "/rest/service-types/chart",
    },
    Object {
      "id": "container",
      "name": "Container",
      "uri": "/rest/service-types/container",
    },
    Object {
      "id": "vm",
      "name": "Virtual Machine",
      "uri": "/rest/service-types/vm",
    },
    Object {
      "id": "zone",
      "name": "Kubernetes Cluster",
      "uri": "/rest/service-types/zone",
    },
  ],
  "start": 0,
  "total": 4,
}
`;

exports[`basic Session 1`] = `
Array [
  "token",
  "userUri",
]
`;

exports[`basic Status 1`] = `
Array [
  "service",
  "database",
]
`;

exports[`basic Tag 1`] = `
Array [
  "id",
  "tagKeyUri",
  "name",
  "uri",
]
`;

exports[`basic TagKey 1`] = `
Array [
  "id",
  "name",
  "uri",
]
`;

exports[`basic User 1`] = `
Array [
  "id",
  "email",
  "name",
  "uri",
  "role",
]
`;

exports[`basic VirtualMachineProfile 1`] = `
Array [
  "id",
  "name",
  "description",
  "regionUri",
  "zoneUri",
  "uri",
  "memorySizeGB",
  "cpuCount",
  "diskSizeGB",
  "created",
  "modified",
]
`;

exports[`basic Zone 1`] = `
Array [
  "id",
  "name",
  "uri",
  "metrics",
  "regionUri",
  "zoneTypeUri",
  "providerUri",
  "status",
  "state",
  "currentTasks",
  "clusters",
  "managed",
  "resourceCapacity",
  "created",
  "modified",
]
`;

exports[`basic ZoneType 1`] = `
Object {
  "count": 3,
  "members": Array [
    Object {
      "id": "aws-az",
      "name": "AWS Availability Zone",
      "uri": "/rest/zone-types/aws-az",
    },
    Object {
      "id": "vcenter",
      "name": "VMware Vcenter Availability Zone",
      "uri": "/rest/zone-types/vcenter",
    },
    Object {
      "id": "k8s",
      "name": "Kubernetes Availability Zone",
      "uri": "/rest/zone-types/kubernetes",
    },
  ],
  "start": 0,
  "total": 3,
}
`;
